Polyadic.Json/NewType/JsonNewTypeConverter.cs

```diff
@@ -0,0 +1,68 @@
+using System;
+using System.Linq.Expressions;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Polyadic.Json.NewType
+{
+    internal sealed class JsonNewTypeConverter<TNewType, TInner> : JsonConverter<TNewType>
+        where TNewType : notnull
+    {
+        private readonly Func<TInner, TNewType> _constructor;
+        private readonly Func<TNewType, TInner> _valueAccessor;
+        private readonly JsonConverter<TInner>? _innerConverter;
+
+        public JsonNewTypeConverter(NewTypeMetadata metadata)
+        {
+            _constructor = Lazy(() => CompileConstructor(metadata));
+            _valueAccessor = Lazy(() => CompileValueAccessor(metadata));
+            _innerConverter = metadata.InnerConverter as JsonConverter<TInner>;
+        }
+
+        public override TNewType? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            var innerValue = _innerConverter is { } converter
```

Comment:
I would extract the innerValue to a Function

Reply:
a685c016266d89c620314b5b16526965f9c61e87